### PR TITLE
1250: rebuild containers in CI when dockerfiles change

### LINKED ISSUE #1250: rebuild containers in CI when dockerfiles change
Currently we always use `docker-compose pull` before building `vt` in CI:
https://github.com/DARMA-tasking/vt/blob/f7ce5215ea4647a22b0fb6c8dbc9527308bf057e/scripts/azure-workflow-template.yml#L96-L109

If there are significant changes in Dockerfiles, manual build and push to dockerhub needs to be performed.

**What Needs to be Done?**
Make sure that CI builds the container (instead of pulling) when PR changes Dockerfiles

### PR DESCRIPTION
**Pull request by cz4rs**
fixes #1250

### CODE REVIEW

**Comment by cz4rs** [comment]
note: 72d5254 tests the mechanism for only one build, update to `azure-workflow-template.yml` is required

**Comment by lifflander** [comment]
This looks great! Good idea on the implementation.

**Comment by cz4rs** [comment]
> This looks great! Good idea on the implementation.

Thanks! This is almost ready, but there are still some things to clarify/consider:
- we only check the content of last PR merged (`develop`) or current PR (branches) to see if dockerfiles were made
- on `develop` this can cause problems if we don't update the images in registry (what if penultimate PR triggered rebuild, but the last one didn't? back to `docker-compose pull`)
- luckily we already have `Push container to registry` task and it should work as intended after this PR is merged

note: **`Push container to registry` doesn't seem to work on `develop` at the moment** - we (always) pull, and then push the same image we've just pulled
- `clang-5.0` image was correctly built and pushed from this branch (see recently added `libomp-dev` on the screenshot below)
image

TODO:
- [ ] we should probably limit pushing to `develop` branch (any others? release branches require different tags probably)
- [ ] all images are built if at least one dockerfile is changed - we could add more fine-grained detection / control, but honestly I don't think that added complexity if worth it (we don't change dockerfiles often)

File: ci/azure/azure-clang-5.0-ubuntu-mpich.yml [hunk 1]
```diff
@@ -40,8 +40,8 @@ variables:
   VT_EXTENDED_TESTS: 1
   VT_UNITY_BUILD: 1
   VT_FCONTEXT: 0
-  VT_USE_OPENMP: 0
```
Comment (lifflander) [inline ci/azure/azure-clang-5.0-ubuntu-mpich.yml:43]: Why is this change included here?

**Review by lifflander — APPROVED** [review]
Looks great. I just had a question about the threading change here. Was that just to test that this is working? Isn't that what the removal of whitespace did?

**Review by jstrzebonski — APPROVED** [review]
Looks good to me.

File: ci/azure/azure-clang-5.0-ubuntu-mpich.yml (hunk 1 above)
Comment (cz4rs) [reply ci/azure/azure-clang-5.0-ubuntu-mpich.yml:43]: whitespace change was indeed enough to test this PR
`VT_USE_OPENMP` here is a leftover of #1233, but I thought I would enable it without creating a separate issue

**Comment by cz4rs** [comment]
ff7031b makes sure that we only push from `develop` branch.
I will merge this in the morning (CET) if the builds are green.

File: ci/azure/azure-clang-5.0-ubuntu-mpich.yml (hunk 1 above)
Comment (lifflander) [reply ci/azure/azure-clang-5.0-ubuntu-mpich.yml:43]: Cool, sounds good to me.
